src/compiler/checker.ts

```diff
@@ -24699,6 +24700,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
         }
 
+        function inferFromMappedProperties(source: Type, target: MappedType) {
+            const properties = getPropertiesOfObjectType(source);
+            for (const sourceProp of properties) {
+                if (!some(sourceProp.declarations, hasSkipDirectInferenceFlag)) {
+                    const sourcePropType = getTypeOfSymbol(sourceProp);
+                    if (!couldContainTypeVariables(sourcePropType)) {

```

Comment:
...? It's `target` that has our potential target type variable in it, so why are we checking `sourcePropType` for variable-possibility here? This prevents, eg, inferring from `string` to `T`, while allowing `U` to `T`. It's weird.

Reply:
This was a poor attempt to avoid recursing into circular structures in the [mappedTypeRecursiveInference](
https://github.com/microsoft/TypeScript/blob/7f292bf2a19aa14ed69a55e646111af9533d8f1c/tests/cases/compiler/mappedTypeRecursiveInference.ts) test case. 

I fixed this now with `invokeOnce` but that creates some other problems. I'll have to investigate this further.